### PR TITLE
Add inline syntax to override attributes/associations.

### PR DESCRIPTION
**Pull request by beauby**
This PR brings the following syntax for overriding attributes/associations and declaring virtual attributes/associations in a serializer:
```ruby
class PostSerializer < ActiveModel::Serializer
  attribute :name do
    "#{object.first_name} #{object.last_name}"
  end

  has_many :comments do
    object.comments.limit(10)
  end
end
```
that is currently equivalent to the usual
```ruby
class PostSerializer < ActiveModel::Serializer
  attribute :name
  def name
    "#{object.first_name} #{object.last_name}"
  end

  has_many :comments
  def comments
    object.comments.limit(10)
  end
end
```
Note that this change is not only cosmetic, as it is a prerequisite to move away from the old way (defining methods on the serializer, which caused many issues when the attribute name was that of a reserved method name (ref #1261)), and refactor the way attributes are handled in a safer way.

Update: added overriding of associations as well + factoring out of the attributes logic.

### CODE REVIEW

**Comment by bf4** [comment]
Nice notion!  But it increases complexity for a solved problem.  If you want to prove the problem isn't solved, and this fixes it, use attributes named 'test' and 'select' that fail before the change and work after.

**Comment by beauby** [comment]
@bf4 You're right, in the current state of things, this does not solve the aforementioned problem. However, in order to solve it, we need to get rid of users defining a virtual/overridden attribute as a method on the serializer, which this PR offers a way to. 
The next PR in this series will extract attributes handling into a mixin and store the attributes' values into a hash instead of defining a method on the serializer.

**Comment by bf4** [comment]
I still don't see how this helps in any way.  Show me! ;)

B mobile phone

**Comment by beauby** [comment]
@bf4 See #1263. Once we remove the `public_send` in `attribute_value`, no more clash with existing methods.

**Comment by bf4** [comment]
Well, I still want to hold on this

B mobile phone

**Comment by beauby** [comment]
@bf4 Sure, let's talk about it when you're back.

File: test/serializers/attribute_test.rb [hunk 1]
```diff
@@ -71,6 +71,21 @@ def id
...
+
+      PostWithVirtualAttribute = Class.new(::Model)
+      class PostWithVirtualAttributeSerializer < ActiveModel::Serializer
+        attribute :name do
+          "#{object.first_name} #{object.last_name}"
+        end
+      end
+
+      def test_virtual_attribute_block
+        post = PostWithVirtualAttribute.new(first_name: 'Lucas', last_name: 'Hosseini')
```
Comment (bf4) [inline test/serializers/attribute_test.rb]: in this case, you might want to test `post.attributes` since that's the immediate case.. tests less layers

Comment (beauby) [inline test/serializers/attribute_test.rb]: Agreed

File: lib/active_model/serializer/attributes.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+        def attributes(*attrs)
+          attrs = attrs.first if attrs.first.class == Array
+
+          attrs.each do |attr|
+            attribute(attr)
+          end
+        end
+
+        def attribute(attr, options = {}, &block)
```
Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: Added the `&block` param.

File: lib/active_model/serializer/attributes.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          key = options.fetch(:key, attr)
+          _attributes_keys[attr] = { key: key } if key != attr
+          _attributes << key unless _attributes.include?(key)
+
+          ActiveModelSerializers.silence_warnings do
+            define_method key do
+              if block_given?
+                instance_eval(&block)
+              else
```
Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: Added the `if block_given?; instance_eval(&block); else` clause.

File: lib/active_model/serializer/attributes.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
+module ActiveModel
+  class Serializer
+    # Defines the attributes to be serialized.
+    #
+    module Attributes
```
Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: This module is directly extracted from the serializer, no modifications to the code have been made, except for:
 1. the definition of an `inherit_attributes` method, rather than overriding `inherited` as in the `Associations` module, which made it impossible for the two modules to define the inherited properties
 2. the optional block parameter to the `attribute` method.

File: lib/active_model/serializer/associations.rb [hunk 5]
```diff
@@ -28,8 +27,8 @@ class << base
       end
 
       module ClassMethods
-        def inherited(base)
-          base._reflections = self._reflections.try(:dup) || []
+        def inherit_associations(base)
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Overriding the `inherited` method does not work when more than one modules do it.

File: lib/active_model/serializer.rb [hunk 6]
```diff
@@ -57,9 +56,9 @@ def self.digest_caller_file(caller_line)
     end
 
     def self.inherited(base)
+      inherit_attributes(base)
+      inherit_associations(base)
```
Comment (bf4) [inline lib/active_model/serializer.rb]: super weird code tells us something isn't right in the design

File: lib/active_model/serializer.rb [hunk 7]
```diff
@@ -168,18 +147,6 @@ def json_key
...
-
-      attributes.each_with_object({}) do |name, hash|
-        if self.class._fragmented
-          hash[name] = self.class._fragmented.public_send(name)
-        else
-          hash[name] = send(name)
-        end
-      end
-    end
-
```
Comment (bf4) [inline lib/active_model/serializer.rb]: funny, I see this as **the** fundamental concern of the serializer and probably the only thing I **wouldn't** remove (though maybe I'd extract method object)

File: lib/active_model/serializer/associations.rb (hunk 5 above)
Comment (bf4) [inline lib/active_model/serializer/associations.rb]: sounds like this should be a bug fix, not part of a feature

**Comment by bf4** [comment]
I'm still very :-1: on this. Sorry.

File: lib/active_model/serializer.rb (hunk 6 above)
Comment (beauby) [inline lib/active_model/serializer.rb]: What wasn't right in the design was that the `Associations` module was overriding the `inherited` method of the `Serializer` class, which worked fine as long as only one module was doing that. I'm all up for suggestion as to making this cleaner though.

File: lib/active_model/serializer.rb (hunk 7 above)
Comment (beauby) [inline lib/active_model/serializer.rb]: I personally don't think associations are any less fundamental to serializers than attributes. Plus extracting semantically delimited code to a module does not make it less fundamental in my opinion.

File: lib/active_model/serializer/associations.rb (hunk 5 above)
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Agreed. I just happened to stumble upon this problem while writing this PR.

**Comment by beauby** [comment]
@bf4 Any specific reason?

**Comment by bf4** [comment]
I don't think it's good for AMS.

**Comment by beauby** [comment]
Would you care to give
> any specific reason?

**Comment by beauby** [comment]
Because as you can imagine, I *do* think it's good for AMS, so it's kind of a Mexican standout.

**Comment by bf4** [comment]
except others are welcome to show up and approve :)

File: lib/active_model/serializer/associations.rb (hunk 5 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: I'm always ok with small bug fixes here and there. Esp if they are stumbled upon, and aren't documented anywhere. Better to catch it early before it becomes an actual problem.

File: test/serializers/attribute_test.rb (hunk 1 above)
Comment (NullVoxPopuli) [inline test/serializers/attribute_test.rb]: I'd like to see this too.

File: lib/active_model/serializer/attributes.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+          end
+        end
+
+        def attribute(attr, options = {}, &block)
+          key = options.fetch(:key, attr)
+          _attributes_keys[attr] = { key: key } if key != attr
+          _attributes << key unless _attributes.include?(key)
+
+          ActiveModelSerializers.silence_warnings do
+            define_method key do
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: this absolutely needs to be document above the method signature with some examples. 
I am a HUUUGE fan of metaprogramming, but it can be a pain in the butt if people aren't expecting it. So, having some example syntax in a method yardoc would be fabulous.

**Comment by NullVoxPopuli** [comment]
ok, so --

I am a fan of, especially for the json api, how there are numerous things that can be configured via blocks. I think that's very consistent, and doesn't leave our users guessing as to which syntax to use.

I do agree that the way module inheritance has been going has been kinda sketchy, but I believe that is a bigger overall architecture discussion for another time.

@beauby, you have my :+1: 

@bf4, do have any specific concerns about this for AMS?

I do know some static analysis tools will complain if there is too much code at the class level, and prefer things to be done in methods (as they are more easily testable, stubbable, greppable etc)

File: lib/active_model/serializer/attributes.rb (hunk 8 above)
Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: Agreed, although this code is already in AMS. I just added the `if block_given?; ...; else`.

Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: Never hurts to add documentation. Code pre-existing isn't an excuse to *not* add docs. :-\

Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: Sure, as I said, I agree. I was just making clear that I didn't just throw loads of metaprogramming magic without comments.

Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: oh :-) :+1:

Comment (bf4) [inline lib/active_model/serializer/attributes.rb]: 1) we should be conservative in adding new interfaces because then we need to support it
2) if we need this feaure, why? If we don't need it, what are the trade offs of adding it? Not adding it?
3) how does it affect ease of understand usage? Ease of maintenaince? Complexity of either?
4) is it extending existing behavior or modifying it? Could it bd done another way?
5) how do you see this feature evolving?  What are its limitations?  Can it be easily customized? Eg adapter no longer need to be in thd ams namespace or even inherit the base.  So, it's easy to extend (though the boundaries could be clearer and the interfaces more consitent.  Am::s should probably be called a resource and implement the serializers::json interface
(Writing in phone; apologies for anything unclear)
>

**Comment by beauby** [comment]
@bf4:
> 1) we should be conservative in adding new interfaces because then we need to support it

I agree. We should also provide the best possible interfaces to our users.

> 2) if we need this feaure, why? If we don't need it, what are the trade offs of adding it? Not adding it?

This feature has two main benefits:
 + When overriding attributes/associations (or declaring virtual attributes/associations), it forces the user to have the redefinition next to the definition, thus encouraging code clarity.
 + If we want to move away from dynamically declaring methods on the serializer in order to access the object's properties (which causes clashes with existing method names, and hard to understand errors) , we need to provide a different way to override attributes/associations (which this feature does).

> 3) how does it affect ease of understand usage? Ease of maintenaince? Complexity of either?

As stated above, I believe it improves ease of usage. Complexity and ease of maintenance are hardly increased.

> 4) is it extending existing behavior or modifying it? Could it bd done another way?

Currently, it is extending existing behavior. It could be done in an other way.

> 5) how do you see this feature evolving?  What are its limitations?  Can it be easily customized? 

I'm not sure about what you mean by evolving, limitations and customization in this setting.

> Eg adapter no longer need to be in thd ams namespace or even inherit the base.  So, it's easy to extend (though the boundaries could be clearer and the interfaces more consitent. 

Not sure I understand. This PR does not deal with adapters in any way.

> Am::s should probably be called a resource 

Sounds sensible

> and implement the serializers::json interface

The `ActiveModel::Serializers::JSON` interface is just `as_json/from_json`. What do you mean?

File: lib/active_model/serializer/attributes.rb (hunk 4 above)
Comment (beauby) [inline lib/active_model/serializer/attributes.rb]: Fixed [the situation](https://github.com/beauby/active_model_serializers/commit/af8fc7e80ed00f67df68ba0f5ad15efcd78d2508). Reverted to overriding `inherited` at module-level, but calling `super` each time in order to decorate the serializers.

**Comment by joaomdmoura** [comment]
Hey everyone, first of all, great work @beauby.
I understand @bf4's point.

We are not solving "defining methods on the serializer" problem **yet**, IMO the main achievement here is aesthetics. Don't take me wrong, it's good aesthetics, and I like it, in favor of it :smile:. 

The concern that I share with @bf4 is,**if** there is a trade-off involved and how big is it. Any piece of code might hunt us down later, even if it seems simple, that's why I think @bf4 made the following question:

> if we need this feature, why?

No. But it's a nice feature, and I agree with @beauby:

> When overriding attributes/associations (or declaring virtual attributes/associations), it forces the user to have the redefinition next to the definition, thus encouraging code clarity.

I'd say to hold overall small/medium features like this one, and focus on the `mandatory` ones first.

About this feature itself, it LGTM, we could merge it now or later after mandatory stuff, but would be nice to know more about this @bf4:

> I don't think it's good for AMS.

Maybe we can talk about it over slack and avoid a huge thread in here :smile:

**Comment by bf4** [comment]
So, I think I would be more :+1: on this DSL if it were optional. i.e. 

Given

```
class UserSerializer < ActiveModel::Serializer
  attribute :testee, key: :test
end
```
right now the `attribute` macro does one thing and expects another:

1. It defines an instance method `test` on the UserSerializer (if not already defined, hence the bug for e.g. Kernel#test)
2. When the adapter calls `user_serializer.attributes` it calls `:testee` on the serializer or fragment cache

The UserSerializer also makes available the the model it wraps as `object` so that devs can define their own methods for the attributes to refer to.  (serialization_scope is another method that provides an instance method, current_user. We'll be adding a serialization_context method soon, too, as well.)

The problem with this approach is that it requires us to be vigilant not to define instance methods on ActiveModel::Serializer that might also be attributes

So, the general need is to avoid this conflict.

One solution would be for all AMS methods to be prefixed with something like `_` to minimize the chances of collision and maybe have a blacklist for undesired methods.

The type of solution in this PR is reasonable, in that it can solve the problem by removing the need to define methods on the serializer instance at all.  

Every attribute declaration would now define an attribute object that the class keeps track of.  This object may be a method object or a Proc.  So then, when we calls attributes, we call one of these method objects.  We could no longer override or define methods on the instance, but we could define them inline in a Proc/Lambda/Method object etc.

This would be a big breaking change but could have big benefits, depending on how it's implemented.

Now, we can exchange the innards of `Serializer.attribute(attr, options = {})`

```ruby
          key = options.fetch(:key, attr)
          _attributes_keys[attr] = { key: key } if key != attr
          _attributes << key unless _attributes.include?(key)
          define_method :test do
              object.read_attribute_for_serialization(:testee)
            end
          end unless method_defined?(:test) || _fragmented.respond_to?(:testee)
```

for

```ruby
          key = options.fetch(:key, attr, &block)
          _attributes[key] = block || proc do 
            object.read_attribute_for_serialization(:testee)
          end
```

which is then used in `Serializer#attributes` similarly

```ruby
      def attributes
        attributes = self.class._attributes.dup

        attributes.each_with_object({}) do |name, hash|
          if self.class._fragmented
            hash[name] = self.class._fragmented.public_send(name)
          else
            hash[name] = send(name)
          end
        end
      end
```

becomes

```
      def attributes
        self.class._attributes.each_with_object({}) do |key, attributes|
          if self.class._fragmented
            attributes[key] = self.class._fragmented.public_send(key)
          else
            attributes[key] = _attributes.call
          end
        end
      end
```

And maybe we can preserve the interface by introducing an `Attribute = Struct.new(:callable) do def call; callable.call; end; end`

Sot that we can define the attribute method as 

```ruby
_attributes[:test] = Attribute.new(block || ->{  object.read_attribute_for_serialization(:testee) })
```

and call it as

```ruby
  include ActiveModel::Serializers::JSON
  def read_attribute_for_serialization(key)
    self.class._fragmented ? _self.class._fragmented.public_send(key) : self.class._attributes[key].call
  end
  def attributes
    @attributes ||= self.class.attributes.keys.each_with_object({}) do |key, attributes|
      attributes[key] = read_attribute_for_serialization(key)
    end
  end
```

which is super nice because the our `Serializer` implements the [ActiveModel::Serialization](https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/serialization.rb) ([JSON](https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/serializers/json.rb)) interface and is the basically a serialization formatter for the the record. ref: https://github.com/rails-api/active_model_serializers/pull/1301 and https://github.com/rails-api/active_model_serializers/issues/1298

So, a dev that wanted to define a custom method on the Serializer would either

```ruby
attribute :test, -> { object.testee }
# or
attribute :testee, key: :test
# or
_attributes[:test] = Attribute.new(->{ object.read_attribute_for_serialization(:testee) })
```

but that's not quite what is in this PR and doesn't require the concerning :kissing_smiling_eyes: removal of attributes methods to a module since they are now even more fundamental to the class.

:)

(note: written without proofreading, so some details might not mesh quite right, but should still make sense.)

**Comment by beauby** [comment]
That's very close to the following PR I mentioned at the beginning of the conv (except it handles only blocks). I fail to see what you don't feel confortable with here (except for extracting Attributes into its own module, which we could debate about in an other PR).

**Comment by bf4** [comment]
like I said, it's pretty similar, but I don't like passing the block all over the place.  I'd like a more surgical change.

**Comment by bf4** [comment]
Also, what I described solves a problem. It's more than just sugar

**Comment by beauby** [comment]
@bf4 As I stated in the PR description, this is only part 1 of a 2 part PR:
> Note that this change is not only cosmetic, as it is a prerequisite to move away from the old way (defining methods on the serializer, which caused many issues when the attribute name was that of a reserved method name (ref #1261)), and refactor the way attributes are handled in a safer way.

Plus, at the end of the day, this PR is passing a proc around, not a block.

**Comment by beauby** [comment]
So to sum up my vision of things here:
 + I'd like to introduce this inline block syntax to override attributes and associations (and declare virtual ones)
 + Because it promotes cleaner serializers for the users (all the logic for one attribute/association kept in one place of the file)
 + And also so that we can remove this nasty dynamic method generation for each attribute/association, which leads to very surprising errors (c.f. mentioned issues)
 + So that we're left with clean serializer classes, on which all the methods defined are AMS ones, and possibly some user-defined which could be used for
 + conditionally including/excluding attributes/associations (the proposed `include_if` / `exclude_if` options, which would only take symbols (referring to user-defined methods), no lambdas)
 + conditionally chosing the serializer for an association (the proposed extension of the `serializer` options which could take a String, a Constant, or a *Symbol* (referring to a user-defined method).

**Comment by bf4** [comment]
>  passing a proc around, not a block.

I'm not sure I understand the significance of the distinction. I don't follow. It's a callable object. Proc, Lambda, Method are all objects that are all callable.

**Comment by beauby** [comment]
You said 
>  I don't like passing the block all over the place

I said 
> this PR is passing a proc around, not a block.

**Comment by bf4** [comment]
In any case, I think the change in this PR could be much smaller, solve the problem ( no need for #1263 ), and actually improve the design and clarity of the code, as I discussed in https://github.com/rails-api/active_model_serializers/pull/1262#issuecomment-151545353

Still needs work, IMHO. Let's focus this PR on defining/encapsulating attribute conversions at the class level.  Let me know if you want to pair.

**Comment by bf4** [comment]
Where are we on this? want me to submit a PR to your branch?

**Comment by beauby** [comment]
Will update soon.

**Comment by bf4** [comment]
awesomes

**Comment by beauby** [comment]
@bf4 See 1311 which covers half of this PR (the associations part).

File: lib/active_model/serializer/associations.rb [hunk 9]
```diff
@@ -13,9 +13,8 @@ module Associations
       DEFAULT_INCLUDE_TREE = ActiveModel::Serializer::IncludeTree.from_string('*')
 
       included do |base|
-        class << base
-          attr_accessor :_reflections
-        end
+        base.class_attribute :_reflections
```
Comment (bf4) [inline lib/active_model/serializer/associations.rb]: did you intend to allow instance reader/writers? did you intend to make it inheritable?

File: lib/active_model/serializer.rb [hunk 10]
```diff
@@ -83,37 +79,6 @@ def self.type(type)
...
-    # @example
-    #   class AdminAuthorSerializer < ActiveModel::Serializer
-    #     attributes :id, :name, :recent_edits
-    def self.attributes(*attrs)
-      attrs = attrs.first if attrs.first.class == Array
-
-      attrs.each do |attr|
-        attribute(attr)
-      end
-    end
```
Comment (bf4) [inline lib/active_model/serializer.rb]: wrote a bunch but github seems to have lost it when I clicked on the tab to get the link to https://github.com/rails-api/active_model_serializers/pull/1262#issuecomment-151545353 re: a latent Attribute class and not removing the attributes api definition from the serializer or changing the behavior of _reflections or _attributes or _attribute_keys 

gonna tweet about it.

tl;dr I'd rather discuss adding a new superclass (the Attributes mxin) separately from adding new behavior (removing the object instance method in favor of defining attribute and association mappings in the attribute / association declarations)

Comment (beauby) [inline lib/active_model/serializer.rb]: You'll be delighted to know that I'm doing exactly that as we speak. As said, this is only half the other PR (only the association related stuff).

Comment (bf4) [inline lib/active_model/serializer.rb]: :heart: :smile: :heart:

**Comment by bf4** [comment]
@beauby I put together something somewhat minimal based on your work here https://github.com/rails-api/active_model_serializers/pull/1356 (which includes your commits)

**Comment by bf4** [comment]
Closed by https://github.com/rails-api/active_model_serializers/issues/1356
